File: 3_Signalanalyse/Python/Aliasing.py
import numpy as np
import matplotlib.pyplot as plt
import scipy.io.wavfile as sio

# read wav file
[f_s, y] = sio.read('r2d2whereareyou.wav')

n = np.size(y)
t = np.linspace(0, n/f_s, n)

plt.figure()
plt.plot(t, y)
plt.xlabel('Time $t$ / s')
plt.title('original signal')

delta_f = f_s/n
f = np.arange(0, f_s, delta_f)

plt.figure()
ft = abs(np.fft.fft(y))
plt.plot(f, ft)
plt.xlabel('frequency $f$ / Hz')
plt.title('spectrum of original signal')

# downsample signal
p = 2               # downsampling factor
f_s_new = f_s/p     # new sampling frequency
y_downsample = y[0::p]
n_new = np.size(y_downsample)
t = np.linspace(0, n_new/f_s_new, n_new)

plt.figure()
plt.plot(t, y_downsample)
plt.xlabel('time $t$ / s')
plt.title('signal sampled with '+str(f_s_new)+' Hz')

delta_f_new = f_s_new/n_new
f_new = np.arange(0, f_s_new, delta_f_new)

plt.figure()
plt.plot(f_new, abs(np.fft.fft(y_downsample)))
plt.xlabel('frequency $f$ / Hz')
plt.title('spectrum of signal sampled with '+str(f_s_new)+' Hz')

# write downsampled signal to .wav
sio.write('downsample.wav', int(f_s_new), y_downsample)
plt.show()

File: 3_Signalanalyse/MATLAB/Aliasing.m
clear all
close all
clc

%% load and plot
% read file
[y, sampl_freq] = audioread('r2d2whereareyou.wav');

n = length(y);                   % signal length
delta_t = 1/sampl_freq;
t = 0:delta_t:(n-1)/sampl_freq;  % time vector

% plot signal
figure
plot(t, y)
xlabel('Time {\itt} / s')
ylabel('Original singal')

% play signal
sound(y, sampl_freq)

% spectrum
delta_f = sampl_freq/n;
f = 0:delta_f:sampl_freq-delta_f;
ft = abs(fft(y));

figure
plot(f, ft)
xlabel('frequency {\itf} / Hz')
ylabel('spectrum of original signal')


%% downsample
p = 2;      % downsampling factor
sampl_freq_downsample = sampl_freq/p;   % new downsampling frequency
y_downsample = downsample(y, p);
n_downsample = length(y_downsample);
t = 0:1/sampl_freq_downsample:(n_downsample-1)/sampl_freq_downsample;

figure
plot(t, y_downsample, 'r')
xlabel('Time {\itt} / s')
ylabel(['signal sampled with ', num2str(sampl_freq_downsample), ' Hz)'])

delta_f_downsample = sampl_freq_downsample/n_downsample;
f_downsample = 0:delta_f_downsample:sampl_freq_downsample-delta_f_downsample;
figure
plot(f_downsample, abs(fft(y_downsample)), 'r')
xlabel('Frequency {\itf} / Hz')
ylabel(['spectrum of signal samples with ', num2str(sampl_freq_downsample), ' Hz)'])

sound(y_downsample, sampl_freq_downsample)
% audiowrite('downsample.wav', y_downsample, sampl_freq_downsample);
